source/common/config/new_grpc_mux_impl.cc

```diff
@@ -0,0 +1,238 @@
+#include "common/config/new_grpc_mux_impl.h"
+
+#include "common/common/assert.h"
+#include "common/common/backoff_strategy.h"
+#include "common/common/token_bucket_impl.h"
+#include "common/config/utility.h"
+#include "common/protobuf/protobuf.h"
+#include "common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Config {
+
+NewGrpcMuxImpl::NewGrpcMuxImpl(Grpc::RawAsyncClientPtr&& async_client,
+                               Event::Dispatcher& dispatcher,
+                               const Protobuf::MethodDescriptor& service_method,
+                               Runtime::RandomGenerator& random, Stats::Scope& scope,
+                               const RateLimitSettings& rate_limit_settings,
+                               const LocalInfo::LocalInfo& local_info)
+    : dispatcher_(dispatcher), local_info_(local_info),
+      grpc_stream_(this, std::move(async_client), service_method, random, dispatcher, scope,
+                   rate_limit_settings) {}
+
+Watch* NewGrpcMuxImpl::addOrUpdateWatch(const std::string& type_url, Watch* watch,
+                                        const std::set<std::string>& resources,
+                                        SubscriptionCallbacks& callbacks,
+                                        std::chrono::milliseconds init_fetch_timeout) {
+  if (watch == nullptr) {
+    return addWatch(type_url, resources, callbacks, init_fetch_timeout);
+  } else {
+    updateWatch(type_url, watch, resources);
+    return watch;
+  }
+}
+
+void NewGrpcMuxImpl::removeWatch(const std::string& type_url, Watch* watch) {
+  updateWatch(type_url, watch, {});
+  auto entry = subscriptions_.find(type_url);
+  if (entry == subscriptions_.end()) {
+    ENVOY_LOG(error, "removeWatch() called for non-existent subscription {}.", type_url);
+    return;
+  }
+  entry->second->watch_map_.removeWatch(watch);
+}
+
+void NewGrpcMuxImpl::pause(const std::string& type_url) { pausable_ack_queue_.pause(type_url); }
+
+void NewGrpcMuxImpl::resume(const std::string& type_url) {
+  pausable_ack_queue_.resume(type_url);
+  trySendDiscoveryRequests();
+}
+
+bool NewGrpcMuxImpl::paused(const std::string& type_url) const {
+  return pausable_ack_queue_.paused(type_url);
+}
+
+void NewGrpcMuxImpl::onDiscoveryResponse(
+    std::unique_ptr<envoy::api::v2::DeltaDiscoveryResponse>&& message) {
+  ENVOY_LOG(debug, "Received DeltaDiscoveryResponse for {} at version {}", message->type_url(),
+            message->system_version_info());
+  auto sub = subscriptions_.find(message->type_url());
+  if (sub == subscriptions_.end()) {
+    ENVOY_LOG(warn,
```

Comment:
No coverage here either.

Reply:
Here is the one instance I will leave as not an assert, since it's about the server's behavior.

Added a test, in new_grpc_mux_impl_test (temporary, will be merged back into grpc_mux_impl_test upon unification).